### ISSUE TITLE
Better way to identify columns for visualization settings

### ISSUE THREAD

**Issue by tlrobinson** [metabase/metabase#11829]
## Background

We need a way to key column-level `visualization_settings` to query result columns in stable way as the query changes.

## Requirements

The settings should be preserved when manipulating the query in certain ways, e.x.

1. adding/removing aggregations/breakouts/fields
2. adding/removing nesting, which changes `field_ref`s from `field-id` etc to `field-literal`
3. converting a GUI question to SQL (do we have any option besides `name`?)

We've changed how this works over time for different reasons:

1. Originally we used `column.name`, but fields with the same name (e.x. aggregations of the same type) would be suffixed with an index like `_2`
    * e.x. if you have "sum of foo" and "sum of bar", then remove "sum of foo", the existing "sum of foo" column settings would be applied to "sum of "bar")
2. Then we used a field reference, initially computed from column metadata like `id`, `expression-name`, etc. but this was error prone and similar logic was duplicated https://github.com/metabase/metabase/issues/8859
3. Currently using the recently added `field_ref`

## Possible solutions

(these are properties on the column objects)

1. `name` as is
    - doesn't preserve settings when multiple columns with the same `name` are added/removed/reordered because duplicate names are suffixed with a position-based index
2. `name` but with more verbose names (e.x. `sum_of_total` and `category_via_product_id`) that are unlikely to be duplicated
3. `field_ref`
    - doesn't preserve settings when switching between non-nested (`:field-id` etc) and nested (`:field-literal`) queries 
    - suffers same issue as using `name` if there are duplicate columns in `source-query`
4. `key` with something like a more unique `name` as described in # 2
5. `key` that uses a unique identifier stored in MBQL clauses (generated by FE when adding a clause) in an "options" argument, e.x. `[:sum [:field-id 1] {:key "207c67f6" }]` (also a good place to add `:display-name` for notebook's step naming concept)
    - would need to fallback to something else for native queries and old queries without `key`. 
6. `key` with the serialized MBQL (similar to solution 4)
7. Use `name` but add logic to the frontend to update settings keys as field `name`s change (this is likely madness)

Here's a matrix of which cases these solutions handle: https://docs.google.com/spreadsheets/d/1L0KDEyxR5FMlKAoorDjQ2KELLRetao0RuQNSVQkY4b8/edit#gid=0

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by tlrobinson** [comment]
I'm in favor of something like 3 or 5.

Ultimately I think solution 5, adding unique identifiers to MBQL clauses and including it on column objects (including downstream nested query columns), would be the most robust. This would handle:

1. adding/removing/reordering columns that happen to have the same name
2. switching between nested/non-nested queries
3. modifying existing clauses' aggregation/field

I think making it so any MBQL clause can end with an optional "options" map would make this easy (and provide a natural place to put notebook step labels)

```edn
{
  :source-table 1,
  :filter [:=, [:field-id, 1], 42, {:key 1, :display-name "The meaning of life" }]
  :aggregation [[:count {:key 2}]]
  :breakout [[:field-id 2 {:key 3}]]
  ...
}
```

(not sure if we'd use incrementing keys or UUIDs or something)

**Comment by tlrobinson** [comment]
One case where 3 and 5 might not work is switching from MBQL to native queries, since there's probably no way to carry over the key into the native query (except encoding it in the column name or something wacky)

I think 2 would be the best way to handle that, however I'm not sure how easy it would be to migrate existing nested questions if the column names change.

**Comment by tlrobinson** [comment]
I've started a spreadsheet with these options and cases where they do/don't work here https://docs.google.com/spreadsheets/d/1L0KDEyxR5FMlKAoorDjQ2KELLRetao0RuQNSVQkY4b8/edit?usp=sharing

**Comment by sbelak** [comment]
Why not use the whole MBQL expression as key: `[:sum [:field-id 1]]` ... which would also work with old cards.

**Comment by tlrobinson** [comment]
> Why not use the whole MBQL expression as key: `[:sum [:field-id 1]]` ... which would also work with old cards.

Yeah, that could work. It has similar tradeoffs to option 2 without the need to migrate nested questions. I'll add it to the spreadsheet.

**Comment by ranquild** [comment]
We no longer use field refs in viz settings
